lact: 0.6.0 -> 0.7.4

cything opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done
Changelog for 0.7.0: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.0
Changelog for 0.7.1: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.1
Changelog for 0.7.2: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.2
Changelog for 0.7.3: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.3

Also added `nix-update-script` to `passthru.updateScript` and tested that it works.

Resolves #374358 

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

cything committed:
lact: 0.6.0 -> 0.7.0

changelog: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.0

also add nix-update-script to passthru.updateScript

Atemu commented on pkgs/by-name/la/lact/package.nix:
```diff
@@ -43,6 +45,7 @@ rustPlatform.buildRustPackage rec {
   checkFlags = [
     # tries and fails to initialize gtk
     "--skip=app::pages::thermals_page::fan_curve_frame::tests::set_get_curve"
+    "--skip=tests::snapshot_everything"
```
Could you report this upstream and add a back reference?

Or at least a comment why this fails.

cything replied:
I can't really understand why the test is failing tbh. Here's some logs:
```
2025-01-18T19:02:36.867269Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:36.867441Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:36.867527Z  INFO lact_daemon::server::handler: initialized radeon controller for GPU 1002:6818-174B:E217-0000:01:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/hd7870/card0/device'
2025-01-18T19:02:36.867809Z  INFO lact_daemon::server::handler: initialized 1 GPUs
test config::tests::parse_doc ... ok
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__hd7870.snap.new
2025-01-18T19:02:36.936453Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:36.936639Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:36.936741Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:679A-1458:229A-0000:01:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/r9-280/card0/device'
2025-01-18T19:02:36.937053Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__r9-280.snap.new
2025-01-18T19:02:36.978342Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:36.978644Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:36.978825Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:731F-1DA2:E409-0000:28:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/rx5700xt/card1/device'
2025-01-18T19:02:36.979254Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__rx5700xt.snap.new
2025-01-18T19:02:37.025115Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.025401Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.025600Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:67DF-1DA2:E387-0000:09:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/rx580/card0/device'
2025-01-18T19:02:37.026048Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__rx580.snap.new
2025-01-18T19:02:37.069399Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.069617Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.069770Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:73FF-1EAE:6505-0000:0b:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/rx6600xt/card0/device'
2025-01-18T19:02:37.070103Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__rx6600xt.snap.new
2025-01-18T19:02:37.114193Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.114360Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.115478Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:73BF-1DA2:440E-0000:0c:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/rx6900xt/card0/device'
2025-01-18T19:02:37.116093Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__rx6900xt.snap.new
2025-01-18T19:02:37.173972Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.174187Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.174347Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:7480-F111:0007-0000:03:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/rx7700s/card1/device'
2025-01-18T19:02:37.174624Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__rx7700s.snap.new
2025-01-18T19:02:37.216746Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.216892Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.216970Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:744C-1DA2:E471-0000:03:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/rx7900xt/card0/device'
2025-01-18T19:02:37.217349Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__rx7900xt.snap.new
2025-01-18T19:02:37.261853Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.262023Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.262174Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:163F-1002:0123-0000:04:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/vangogh/card0/device'
2025-01-18T19:02:37.262446Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__vangogh.snap.new
2025-01-18T19:02:37.303623Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.303793Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.303895Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:687F-1043:0555-0000:0b:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/amd/vega56/card0/device'
2025-01-18T19:02:37.304391Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__amd__vega56.snap.new
2025-01-18T19:02:37.347324Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.347453Z ERROR lact_daemon::server::handler: failed to initialize Intel DRM: libdrm_intel.so.1: cannot open shared object file: No such file or directory
2025-01-18T19:02:37.347469Z ERROR lact_daemon::server::gpu_controller: Intel DRM library missing, Intel controls will not be available
2025-01-18T19:02:37.347494Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.347602Z  INFO lact_daemon::server::handler: initialized i915 controller for GPU 8086:56A5-1849:6004-0000:0b:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/intel/a380-i915/card1/device'
2025-01-18T19:02:37.348093Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__intel__a380-i915.snap.new
2025-01-18T19:02:37.390743Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.390871Z ERROR lact_daemon::server::handler: failed to initialize Intel DRM: libdrm_intel.so.1: cannot open shared object file: No such file or directory
2025-01-18T19:02:37.390893Z ERROR lact_daemon::server::gpu_controller: Intel DRM library missing, Intel controls will not be available
2025-01-18T19:02:37.390915Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.390984Z  INFO lact_daemon::server::handler: initialized xe controller for GPU 8086:56A5-1849:6004-0000:0b:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/intel/a380-xe/card0/device'
2025-01-18T19:02:37.391257Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__intel__a380-xe.snap.new
2025-01-18T19:02:37.432986Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.433221Z ERROR lact_daemon::server::handler: failed to initialize Intel DRM: libdrm_intel.so.1: cannot open shared object file: No such file or directory
2025-01-18T19:02:37.433233Z ERROR lact_daemon::server::gpu_controller: Intel DRM library missing, Intel controls will not be available
2025-01-18T19:02:37.433254Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.433299Z  INFO lact_daemon::server::handler: initialized i915 controller for GPU 8086:9B41-17AA:22B1-0000:00:02.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/intel/cometlake/card0/device'
2025-01-18T19:02:37.433575Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__intel__cometlake.snap.new
2025-01-18T19:02:37.474967Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.475134Z ERROR lact_daemon::server::handler: failed to initialize Intel DRM: libdrm_intel.so.1: cannot open shared object file: No such file or directory
2025-01-18T19:02:37.475149Z ERROR lact_daemon::server::gpu_controller: Intel DRM library missing, Intel controls will not be available
2025-01-18T19:02:37.475179Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.475237Z  INFO lact_daemon::server::handler: initialized xe controller for GPU 8086:9A49-17AA:3F19-0000:00:02.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/intel/tigerlake/card0/device'
2025-01-18T19:02:37.475633Z  INFO lact_daemon::server::handler: initialized 1 GPUs
stored new snapshot /home/yt/fun/LACT/lact-daemon/src/tests/snapshots/lact_daemon__tests__intel__tigerlake.snap.new
2025-01-18T19:02:37.516685Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
2025-01-18T19:02:37.516763Z ERROR lact_daemon::server::gpu_controller: NVML is missing, Nvidia controls will not be available
2025-01-18T19:02:37.516816Z ERROR lact_daemon::server::handler: failed to initialize AMDGPU DRM: libdrm.so: cannot open shared object file: No such file or directory, some functionality will be missing
2025-01-18T19:02:37.516875Z  INFO lact_daemon::server::handler: initialized nvidia controller for GPU 10DE:2704-1462:5110-0000:09:00.0 at '/home/yt/fun/LACT/lact-daemon/src/tests/data/nvidia/rtx4080/card1/device'
2025-01-18T19:02:37.517344Z  INFO lact_daemon::server::handler: initialized 1 GPUs
```

Atemu reviewed (approved):
Diff LGTM but I'm currently not near my machine where I use LACT.

Could you described what you tested to be working?

cything commented:
> Could you described what you tested to be working?
@Atemu 
This upgrade was performed by the script. I had to skip the test and add a new hook to get it to pass, but I assume that's expected? Or should I remove it? (if the script is unlikely to work in the future).

Atemu commented:
I meant what parts of the updated program you tried to run and found to be working correctly. Basically I want to ascertain that the program still functions as expected after the update.

cything commented:
@Atemu 
I have a regular integrated GPU and it seems to be working fine as far as I can tell. Working on fixing the test; I think I may have a clue.

![image](https://github.com/user-attachments/assets/f1f5ebc2-c4ef-4578-a9c6-71378d57b939)
![image](https://github.com/user-attachments/assets/62d04962-2893-439e-93c7-14a7fcefb9df)

Atemu commented:
I'd like to see its core functionality such as fan curves or OC to be tested before we merge.

Atemu commented on pkgs/by-name/la/lact/package.nix:
```diff
@@ -63,11 +66,23 @@ rustPlatform.buildRustPackage rec {
-    patchelf $out/bin/.lact-wrapped --add-needed libvulkan.so --add-rpath ${
-      lib.makeLibraryPath [ vulkan-loader ]
+    patchelf $out/bin/.lact-wrapped \
+    --add-needed ${vulkan-loader}/lib/libvulkan.so \
+    --add-needed ${libdrm}/lib/libdrm.so \
+    --add-needed ${libdrm}/lib/libdrm_intel.so \
+    --add-needed ${libdrm}/lib/libdrm_amdgpu.so \
+    --add-needed ${libdrm}/lib/libdrm_radeon.so \
+    --add-needed ${libdrm}/lib/libdrm_nouveau.so \
```
Just add a needed for libdrm.so (like I had for libvulkan.so before) and add it to the library path.

These do not make sense to declare here; they're loaded at runtime.

cything replied:
sorry about that. done.

Atemu commented on pkgs/by-name/la/lact/package.nix:
```diff
@@ -63,11 +66,19 @@ rustPlatform.buildRustPackage rec {
   '';
 
   postFixup = lib.optionalString stdenv.targetPlatform.isElf ''
-    patchelf $out/bin/.lact-wrapped --add-needed libvulkan.so --add-rpath ${
-      lib.makeLibraryPath [ vulkan-loader ]
```
Please keep the logic this way; only adding needed for the .so name, not the exact path to the .so.

AFAIUI, this allows overriding the .so used using LD_LIBRARY_PATH.

cything commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 374771`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>lact</li>
  </ul>
</details>

cything commented:
@Atemu 
I could get the test to pass with:
```
RUSTFLAGS='-C link-arg=-Wl,-rpath,/nix/store/9bfw268g1027wg88b882ljhbjlqmmc2j-vulkan-loader-1.3.296.0/lib -C link-arg=-Wl,-rpath,/nix/store/5s1qvxmp4bgdzi6yq9fqfzizw01jf2mp-libdrm-2.4.124/lib -C link-arg=-Wl,--add-needed,/nix/store/9bfw268g1027wg88b882ljhbjlqmmc2j-vulkan-loader-1.3.296.0/lib/libvulkan.so -C link-arg=-Wl,--add-needed,/nix/store/5s1qvxmp4bgdzi6yq9fqfzizw01jf2mp-libdrm-2.4.124/lib/libdrm.so' cargo insta test
```
(the failing test is an [insta](https://insta.rs/docs/cli/) test so `cargo instal test` to run all insta tests)

But we run `patchelf` after `cargo test` which is why it won't pass in the Nix package. I guess I could get it work by passing `RUSTFLAGS` to `buildRustPackage` then we probably wouldn't need `patchelf` anymore. Let me know what you think.

cything commented:
There's one more reason the test fails. The package depends on [pciid-parser](https://crates.io/crates/pciid-parser) which uses a [hardcoded](https://docs.rs/pciid-parser/0.7.2/src/pciid_parser/lib.rs.html#20) list of paths to lookup PCI database.

cything committed:
lact: fix snapshot_everything test

- use RUSTFLAGS for library paths so that the binary works during
  checkPhase
- patch Database::read() cause it tries to lookup the PCID database in
  /usr/share

cything committed:
lact: 0.7.0 -> 0.7.1

changelog: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.1

cything commented:
Use `substituteInPlace` for the patch and updated to [0.7.1](https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.1)

cything committed:
lact: enable failing intel tests since they pass after the update

cything commented:
Looks like they fixed the tests :)
@Atemu I don't know if you're still interested in reviewing this PR. If not, please excuse the ping.

damiankorcz commented:
There is now a 0.7.2 version with support for RDNA4 GPUs:
https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.2

cything committed:
lact: 0.7.1 -> 0.7.3

changelog: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.2

changelog: https://github.com/ilya-zlobintsev/LACT/releases/tag/v0.7.3

cything commented:
Upgrading hwdata fixed the AMD tests. Thanks @N-Parsons !
@damiankorcz can you please test it out now?

cything commented:
Rebased to resolve conflict

damiankorcz commented:
Hi @cything,

I've overlaid this LACT and hwdata update like so in my `flake.nix`:

In inputs:
```nix
lact-pr.url = "github:cything/nixpkgs?ref=lact";
```

In modules:
```nix
{
    nixpkgs.overlays = [
        (final: prev: {
            lact = final.callPackage "${lact-pr}/pkgs/by-name/la/lact/package.nix" {
                hwdata = final.callPackage "${lact-pr}/pkgs/by-name/hw/hwdata/package.nix" { };
            };
        })
    ];
}
```

This is more a sanity check since I'm not too familiar with overlays. Let me know if that is the expected way to do tests like this.

---
A few issues:

1. LACT complains about a missing daemon on launch, I resolve this by adding the following to my config:
```nix
systemd.services.lact = {
    description = "LACT Daemon";
    after = [ "multi-user.target" ];
    wantedBy = [ "multi-user.target" ];
    serviceConfig = {
        ExecStart = "${pkgs.lact}/bin/lact daemon";
    };
    enable = true;
};
```

Would be nice to have this integrated into the package.

2. Overclocking isn't enabled by default and trying to use the option in the UI gives the following prompt:

![Screenshot_20250317_201748](https://github.com/user-attachments/assets/f5611785-e438-4d00-be46-5ec1819d808f)

Clicking OK and restarting the system, predictably, doesn't change anything.

The solution for me was to set:

```nix
boot.kernelParams = [ "amdgpu.ppfeaturemask=0xfffd7fff" ];
```

I got this from the [NixOS options available for CoreCtrl](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=programs.corectrl). Specifically the `gpuOverclock` ones. We might want to have the same set of options for LACT for convenience.

---

Pretty sure the following issues are something else and not packaging on our end, but regardless:

1. The Information tab does show the correct GPU Model and Card Manufacturer, however it doesn't recognise the Card Model. I specifically have the [Sapphire RX 9070 XT Pulse](https://www.sapphiretech.com/en/consumer/pulse-radeon-rx-9070-xt-16g-gddr6).
![image](https://github.com/user-attachments/assets/0b10b0a4-1f10-4240-ab06-978b9e9772a6)

2. The static Fan control option doesn't work. Switching to it also breaks the other options:
![image](https://github.com/user-attachments/assets/a6d9a9a7-735a-4cba-9265-c2a9f70470cd)

Overall, seems that the Overclocking and Fan control features that are supposed to work are working now. 

Thanks to all for getting this update!

cything commented:
> Would be nice to have this integrated into the package.

A systemd module is outside the scope of  a package. There is however a PR for a module: https://github.com/NixOS/nixpkgs/pull/338059

Thank you so much for testing! and I am really glad that it worked. Hopefully, this can get merged soon.

damiankorcz commented:
Another thing I noticed is this release states:
> The `vulkaninfo` command from `vulkan-tools` is now required. Without it vulkan information will not be displayed.

I do have `vulkan-tools` / `vulkaninfo` available on my system yet the Vulkan information as you can see in the screenshot in my last message isn't being read. Could that be an issue with packaging or do you thing I should ask upstream about this as well?

cything commented:
I also noticed that they say `blueprint-compiler` is no longer needed. Will remove that and add `vulkan-tools` to `buildInputs` once I verify that it still builds.

> I do have vulkan-tools / vulkaninfo available on my system yet the Vulkan information as you can see in the screenshot in my last message isn't being read. Could that be an issue with packaging or do you thing I should ask upstream about this as well?

Not sure if it will make a difference but try adding `vulkan-tools` to `buildInputs` of this package.

damiankorcz commented:
@cything I've tried building with the `lact-test` branch that you made the changes in and it builds fine but still no Vulkan info.

cything commented:
The error I get is `could not load vulkan info: Could not read vulkan info for device`. Looking at the source code, I think this is because the `vulkaninfo --json=<GPU id>` command failed.

@damiankorcz can you try the following?
1. Run `vulkaninfo` and note your GPU id. Should be in a line like so: `GPU id : 0 (Intel(R) UHD Graphics (ICL GT1)):`
2. Run `vulkaninfo --json=<GPU id>`

Also, please post the output of `lact daemon` and `lact` commands.

damiankorcz commented:
@cything I got the latest commit and it seems to have fixed it 🥳
![image](https://github.com/user-attachments/assets/97ac21b7-6a34-4de9-913b-6d60d16df1c3)

cything commented:
> @cything I got the latest commit and it seems to have fixed it 🥳 !

That's great ^^
I guess it didn't work on mine cause I don't have a graphics card.

PedroHLC reviewed (changes requested):
`hwdata` no longer needs that `configureFlags` starting with `0.393`.

cything commented:
> `hwdata` no longer needs that `configureFlags` starting with `0.393`.

Thanks. Fixed!

GaetanLepage reviewed (approved):
Diff LGTM

SuperSandro2000 commented on pkgs/by-name/hw/hwdata/package.nix:
```diff
@@ -6,17 +6,15 @@
   src = fetchFromGitHub {
     owner = "vcrhonek";
     repo = "hwdata";
     rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

SuperSandro2000 commented on pkgs/by-name/la/lact/package.nix:
```diff
@@ -3,43 +3,63 @@
   src = fetchFromGitHub {
     owner = "ilya-zlobintsev";
     repo = "LACT";
     rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

SuperSandro2000 commented:
This PR rebuilds a lot of packages which means we must target staging. Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging) to not potentially ping a lot of people.

---

^ that's due to the hwdata update

cything commented:
Sorry for the delay. I had no idea I need to change the base branch myself. Another PR updating hwdata has been merged to staging, but that one did not remove `configureFlags` which we no longer need so I did not remove my commit.

rthorntn commented:
fyi, 0.7.3 is out and it has an internal card database (and a flatpak), not sure if that changes things.

MerkuR-92 commented:
> Update: [ilya-zlobintsev/LACT#547](https://github.com/ilya-zlobintsev/LACT/issues/547) flatpak for LACT on NixOS has issues.

I'm the one who opened that issue. 

Given how important LACT is as a piece of software for NixOS, and the major (and frequent) changes/bug fixes happening from version to version, I thought it was relevant to find a way to get the Flatpak version working—especially since it's officially maintained and kept up to date by the developer.

Maybe this isn't the right place to bring it up, and it might go against the efforts being made to maintain the package, but I believe that finding a way to fully load the service (with the daemon and its missing components) using the Flatpak as a base could potentially be a better solution.

With this, I've managed to successfully load the daemon:

```
 # LACT Control Daemon (Flatpak)
  systemd.services.lactd = {
    description = "LACT Control Daemon (via Flatpak)";
    after = [ "multi-user.target" ];
    wantedBy = [ "multi-user.target" ];
    serviceConfig = {
      Environment = [
        "FLATPAK_INSTALL_USER=rene"
        "VULKANINFO_COMMAND=${pkgs.flatpak}/bin/flatpak run --filesystem=/tmp --command=vulkaninfo io.github.ilya_zlobintsev.LACT"
      ];
      ExecStart = ''
        ${pkgs.runtimeShell} -c 'eval $(${pkgs.flatpak}/bin/flatpak run --command=lact io.github.ilya_zlobintsev.LACT flatpak generate-daemon-cmd)'
      '';
      Nice = "-10";
      Restart = "on-failure";
    };
  };
```

But as an Nvidia user, there are at least two other components that can't be read/loaded: the PCI ID database and the Nvidia Management Library:

```
[rene@nixos:~]$ systemctl status lactd.service
● lactd.service - LACT Control Daemon (via Flatpak)
     Loaded: loaded (/etc/systemd/system/lactd.service; enabled; preset: ignored)
     Active: active (running) since Fri 2025-04-18 22:52:28 CST; 6h ago
 Invocation: 7e6a8a6318fd447bb030640da3ad1898
   Main PID: 1329 (bash)
         IP: 0B in, 0B out
         IO: 54.3M read, 20K written
      Tasks: 4 (limit: 115449)
     Memory: 61.9M (peak: 64.8M)
        CPU: 193ms
     CGroup: /system.slice/lactd.service
             ├─1329 /nix/store/9nw8b61s8lfdn8fkabxhbz0s775gjhbr-bash-5.2p37/bin/bash -c "eval \$(/nix/store/70ygh183hwz6n7kk1inrj3hlixz6yxnr-flatpak-1.16.0/bin/flatpak run --command=lact io.github.ilya_zlobintsev.LACT flatpak generate-daemon-cmd)"
             └─1358 /var/lib/flatpak/runtime/org.gnome.Platform/x86_64/48/bcc39a760692dffe68b8286a68fb88bf3426862d48e6b11e84b50fd96ac9ea1a/files/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 --library-path /var/lib/flatpak/app/io.github.ilya_zlobintsev.LACT/x86_64/stable/4f89c17b2>

abr 18 22:52:29 nixos bash[1358]: 2025-04-19T04:52:29.054741Z  INFO lact_daemon: service started too early, waiting 8.4 seconds
abr 18 22:52:37 nixos bash[1358]: 2025-04-19T04:52:37.472622Z  INFO lact_daemon::socket: listening on "/run/lactd.sock"
abr 18 22:52:37 nixos bash[1358]: 2025-04-19T04:52:37.472649Z  INFO lact_daemon::server: TCP listener disabled
abr 18 22:52:37 nixos bash[1358]: 2025-04-19T04:52:37.472761Z  WARN lact_daemon::server::handler: could not read PCI ID database: file not found, device information will be limited
abr 18 22:52:37 nixos bash[1358]: 2025-04-19T04:52:37.474038Z ERROR lact_daemon::server::handler: could not load Nvidia management library: a libloading error occurred: libnvidia-ml.so.1: cannot open shared object file: No such file or directory
abr 18 22:52:37 nixos bash[1358]: 2025-04-19T04:52:37.474062Z ERROR lact_daemon::server::gpu_controller: NVML is missing, Nvidia controls will not be available
abr 18 22:52:37 nixos bash[1358]: 2025-04-19T04:52:37.474131Z  INFO lact_daemon::server::handler: initialized nvidia controller for GPU 10DE:2684-1462:5103-0000:01:00.0 at '/sys/class/drm/card1/device'
abr 18 22:52:37 nixos bash[1358]: 2025-04-19T04:52:37.475046Z  INFO lact_daemon::server::handler: initialized 1 GPUs
abr 19 04:55:17 nixos bash[1358]: 2025-04-19T10:55:17.056660Z  WARN lact_daemon::server::gpu_controller::amd: could not load vulkan info: Exit code exit status: 1 for 'vulkaninfo':  error: Opción desconocida --summary
```

And at that point, I’m not sure how to proceed.

cything commented:
I made an attempt at an update. Please test it out and let me know if anything breaks. I dropped the `hwdata` commit cause it caused too many rebuilds making it impossible for me to test anything.

cything commented:
> But as an Nvidia user, there are at least two other components that can't be read/loaded: the PCI ID database and the Nvidia Management Library:

@MerkuR-92  The PCI ID database path is hardcoded so I patched it to use the one from `/nix/store`. Did you even try the 0.7.2 version from this PR? I believe most of the errors from the log you posted were addressed. In any case, please test the recent commit I pushed.

cything committed:
lact: add autoAddDriverRunpath to fix libnvidia-ml.so errors

andi242 commented:
tested the latest changes on my system with a `AMD Radeon RX 7900 XT (0x1002:0x744C:0xCC)`

in the build process the tests for the new 9070xt (folders `lact-daemon/src/tests/data/amd/rx9070xt*` in src) failed.
see  [lact073-build.log](https://github.com/user-attachments/files/19843520/lact073-build.log).
I removed said folders in postPatch and build works for me.

added this to the config for systemd service start
```
  systemd.services.lactd = {
    description = "AMDGPU Control Daemon";
    enable = true;
    serviceConfig = {
      # this path because we don't use pkgs.lact
      ExecStart = "/run/current-system/sw/bin/lact daemon";
    };
    wantedBy = [ "multi-user.target" ];
  };
```

I get one warning on the final build and vulkan detection does not work although `vulkan-tools` is installed on the system: 
```
WARN lact_daemon::server::gpu_controller::amd: could not load vulkan info: Could not run 'vulkaninfo'
```
not sure how to fix that.

cything commented:
> in the build process the tests for the new 9070xt (folders lact-daemon/src/tests/data/amd/rx9070xt* in src) failed.

That's very weird. I cannot reproduce this. The package builds fine on my computer and even on a github actions runner.

andi242 commented:
> That's very weird. I cannot reproduce this. The package builds fine on my computer and even on a github actions runner.

not sure if this is because of the hwdata package. will try and remove cached data and rebuild later today.

cything commented:
> not sure if this is because of the hwdata package. will try and remove cached data and rebuild later today.

This test used to fail when `hwdata` was not up-to-date, but the current nixpkgs version should be fine. Is there something wrong with your `hwdata`?

MerkuR-92 commented:
> @MerkuR-92 The PCI ID database path is hardcoded so I patched it to use the one from `/nix/store`. Did you even try the 0.7.2 version from this PR? I believe most of the errors from the log you posted were addressed (unless it's nvidia specific). In any case, please test the recent commit I pushed.

Pardon my ignorance, but how do I install it?

Because adding `_lact_` to `environment.systemPackages`, I assume it will install the outdated version 0.6.

andi242 commented:
> Pardon my ignorance, but how do I install it?
> 
> Because adding `lact` to `environment.systemPackages`, I assume it will install the outdated version 0.6.

@MerkuR-92  I did that like this:
```
  environment.systemPackages = with pkgs; [
    # ...
    (callPackage ./lact.nix { })
    # ...
];
```
while `lact.nix` is the file containing the derivation in the same folder.
be sure to `git add ` the file when running from a git repo or you'll get a file not found error.

cything commented:
> ```
>   environment.systemPackages = with pkgs; [
>     # ...
>     (callPackage ./lact.nix { })
>     # ...
> ];
> ```

@andi242  You are building against your nixpkgs which is why the test faild (old `hwdata`). If you do it the overlay way it should work cause the flake will use the locked nixpkgs.

andi242 commented:
> @andi242 You are building against your nixpkgs which is why the test faild (old `hwdata`). If you do it the overlay way (the comment I linked above) it should work cause the flake will use the locked nixpkgs.

struggling it out now, thanks for the hint!
I feel a bit dumb because that is a so much more elegant way to do. :smiling_face_with_tear:

cything committed:
lact: add cything as maintainer

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5449

aidalgol commented:
I just tried this on my system by adding `lact` to `systemd.packages` and `environment.systemPackages`.  The lactd service now successfully loads the GPU information, and nvidia management library,
```log
May 17 07:55:52 aidaltower systemd[1]: Started GPU Control Daemon.
May 17 07:55:52 aidaltower lact[1602942]: 2025-05-16T19:55:52.569050Z  INFO lact_daemon::socket: listening on "/run/lactd.sock"
May 17 07:55:52 aidaltower lact[1602942]: 2025-05-16T19:55:52.569069Z  INFO lact_daemon::server: TCP listener disabled
May 17 07:55:52 aidaltower lact[1602942]: 2025-05-16T19:55:52.590839Z  INFO lact_daemon::server::handler: Nvidia management library loaded
May 17 07:55:52 aidaltower lact[1602942]: 2025-05-16T19:55:52.611605Z  INFO lact_daemon::server::handler: initialized nvidia controller for GPU 10DE:2206-1043:87B0-0000:09:00.0 at '/sys/class/drm/card1/device'
May 17 07:55:52 aidaltower lact[1602942]: 2025-05-16T19:55:52.612318Z  INFO lact_daemon::server::handler: initialized 1 GPUs
```
But the LACT GUI is horribly broken.  It looks like this, does not respond to any input, and outputs Vulkan errors from [GSK](https://en.wikipedia.org/wiki/GTK_Scene_Graph_Kit).

This is on `nixos-unstable`, running Plasma 6 Wayland on an RTX3080 with the nvidia proprietary driver version `575.51.02`.

![image](https://github.com/user-attachments/assets/af75047d-5a44-432b-be8a-0df14e023965)
```console
(lact:1605600): Gtk-WARNING **: 07:59:20.578: Unknown key gtk-modules in /home/aidan/.config/gtk-4.0/settings.ini
2025-05-16T19:59:21.022575Z  INFO lact_client::connection::unix: connecting to service at "/run/lactd.sock"

(lact:1605600): Gdk-WARNING **: 07:59:21.575: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (32) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:21.583: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gsk-WARNING **: 07:59:26.193: vkQueueSubmit(): The logical or physical device has been lost. (VK_ERROR_DEVICE_LOST) (-4)

(lact:1605600): Gdk-WARNING **: 07:59:28.572: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:31.118: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:31.119: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:31.119: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:33.061: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:35.385: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:35.488: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gsk-WARNING **: 07:59:36.418: vkQueueSubmit(): The logical or physical device has been lost. (VK_ERROR_DEVICE_LOST) (-4)

(lact:1605600): Gdk-WARNING **: 07:59:37.628: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:37.628: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:37.628: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)

(lact:1605600): Gdk-WARNING **: 07:59:37.762: ../gdk/wayland/gdkcursor-wayland.c:210 cursor image size (64) not an integer multiple of theme size (24)
```

Aleksanaa commented:
> But the LACT GUI is horribly broken. It looks like this, does not respond to any input, and outputs Vulkan errors from [GSK](https://en.wikipedia.org/wiki/GTK_Scene_Graph_Kit). 

I suspect this has something to do with plasma. You seem to be loading adwaita theme onto libadwaita, which I remember wasn't a case for the app itself.

aidalgol commented:
> I suspect this has something to do with plasma. You seem to be loading adwaita theme onto libadwaita, which I remember wasn't a case for the app itself.

I do not have the Adwaita theme configured anywhere, and I don't think the compositor can cause "device has been lost" errors.

andi242 commented:
> > I suspect this has something to do with plasma. You seem to be loading adwaita theme onto libadwaita, which I remember wasn't a case for the app itself.
> 
> I do not have the Adwaita theme configured anywhere, and I don't think the compositor can cause "device has been lost" errors.

Not sure if this issue is something this package is causing. sounds more like upstream is causing it.
It seems like libadwaita is bundled in 0.7.3 and will be dropped in the future.

see here as well: https://github.com/ilya-zlobintsev/LACT/issues/526

JohnRTitor commented:
> But the LACT GUI is horribly broken. It looks like this, does not respond to any input, and outputs Vulkan errors from [GSK](https://en.wikipedia.org/wiki/GTK_Scene_Graph_Kit).

Does this also occur on Lact 6.0?

I tested 7.3 and it seems to be just fine.

```
❯  lact
2025-05-22T16:35:07.794111Z  INFO lact_gui::app: could not connect to socket: Could not connect to daemon: socket file not found
2025-05-22T16:35:07.794133Z  INFO lact_gui::app: using a local daemon
2025-05-22T16:35:07.801183Z  INFO lact_daemon::server::handler: AMDGPU DRM initialized
2025-05-22T16:35:07.801314Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:164E-1458:D000-0000:12:00.0 at '/sys/class/drm/card1/device'
2025-05-22T16:35:07.801479Z  INFO lact_daemon::server::handler: initialized amdgpu controller for GPU 1002:73DF-1DA2:E445-0000:03:00.0 at '/sys/class/drm/card0/device'
2025-05-22T16:35:07.802021Z  INFO lact_daemon::server::handler: initialized 2 GPUs
```

No Vulkan error and it responds to inputs perfectly.

JohnRTitor committed:
lact: switch to finalAttrs, add johnrtitor as maintainer

Signed-off-by: John Titor <50095635+JohnRTitor@users.noreply.github.com>

JohnRTitor committed:
lact: 0.7.3 -> 0.7.4

1. requires ocl-icd now for building
2. no need to skip the test
3. replace some hardcoded commands with nix store paths, which introduces
   dependencies on systemdMinimal and vulkan-tools

Signed-off-by: John Titor <50095635+JohnRTitor@users.noreply.github.com>

JohnRTitor reviewed (approved):
This has been stalled for enough. Tested and seems to work, so I am gonna go ahead and merge this.

Please report with logs and at least configuration information so we can debug your issue with GUI and inputs not working.

GaetanLepage commented:
> Out of curiosity, how long would it take for this to show up in the unstable channel?
> 
> I'm really interested!

Usually 2-3 days (https://grafana.nixos.org/d/LhIq8iLWk/channel-updates?orgId=1&refresh=10s).
You can follow the progress of this PR here: https://nixpkgs-tracker.ocfox.me/?pr=374771

Sly-Harvey commented:
> > > Out of curiosity, how long would it take for this to show up in the unstable channel?
> > > I'm really interested!
> > 
> > 
> > Usually 2-3 days (https://grafana.nixos.org/d/LhIq8iLWk/channel-updates?orgId=1&refresh=10s). You can follow the progress of this PR here: https://nixpkgs-tracker.ocfox.me/?pr=374771
> 
> Thank you!
> 
> Very eager to try it out on my Nvidia GPU.

Yeah i've been waiting on this pr for a very long time so i'm glad it's been merged since this is the only program that i've found with power and fan control that actually works for my gtx 1080

MerkuR-92 commented:
> Yeah, the lact package needs a daemon to be run, there's a pending PR for it
> 
> Which I will merge in a bit
> 
> #411707
> 
> Once it reaches unstable, you can just use the option.

Fantastic, I'll be keeping an eye on it!

Thank you very much.

JohnRTitor commented:
Note that on a Flakes based setup, you can easily import modules from a different branch/channel. 

Example: https://github.com/JohnRTitor/nix-conf/commit/e4a4858e883e869b0918fe87eb131b61bcda6f63

You just need to add another flake input with the master branch (since it's merged to master) and import it like how I did it. 

Then you can use the option normally. You'd need to revert it of course, once it reaches unstable as both option declarations will conflict.

MerkuR-92 commented:
At the moment I'm fine without flakes (although I used it a bit in the past), I'm not really in a hurry to wait right now — at least no more than I already have while hoping for the package update. 😅

Thanks anyway!

Sly-Harvey commented:
> At the moment I'm fine without flakes (although I used it a bit in the past), I'm not really in a hurry to wait right now — at least no more than I already have while hoping for the package update. 😅
> 
> 
> 
> Thanks anyway!

No need to wait, it's possible to setup the daemon manually like I have done [here](https://github.com/Sly-Harvey/NixOS/blob/master/modules/programs/misc/lact/default.nix)

MerkuR-92 commented:
> > At the moment I'm fine without flakes (although I used it a bit in the past), I'm not really in a hurry to wait right now — at least no more than I already have while hoping for the package update. 😅
> > Thanks anyway!
> 
> No need to wait, it's possible to setup the daemon manually like I have done [here](https://github.com/Sly-Harvey/NixOS/blob/master/modules/programs/misc/lact/default.nix) and then lact starts without any issues and I'm able to overclock, power limit, fan curve, etc.

This absolutely works!

Wow, I finally managed to find a solution to adjust these things on Linux (and with a pretty nice interface).

Many thanks to everyone who made this possible for NixOS.

aidalgol commented:
Just for posterity, the issue I had with an earlier version of this PR (https://github.com/NixOS/nixpkgs/pull/374771#issuecomment-2887586278) is gone with the changes merged to `master`.
